GH-38558: [C++] Add support for null sort option per sort key

taepper opened the pull request:
See #38584 for original PR. Will be quoted for this PR description.

### Rationale for this change

> support multi sortkey nulls first.

```
order by i nulls first, j, k nulls first;
```

> The current null sorting only supports all sortkeys, not a certain sortkey, so NullPlacement is extended to the SortKey field. Since the underlying framework is very well written, when modifying this function, you only need to pass the null_placement of each SortKey in. That’s it.

### What changes are included in this PR?

> 1.SortKey structure, NullPlacemnt transfer logic, sorting logic and Ording related, test related
> 2.Substriait related.
> 3.c_glib related.
> 4.SelectK related.
> 5.RankOptions related.

### Are these changes tested?

> yes, I changed the code inside vector_sort_test.cc and performed additional tests.

### Are there any user-facing changes?

> yes, pg database include null sorting of multiple sort keys.

**This PR includes breaking changes to public APIs.** (If there are any breaking changes to public APIs, please explain which changes are breaking. If not, you can remove this.)

I amended the original PR to be less breaking in public APIs.

Still Ordering, SortOptions, RankOptions, and RankQuantileOptions now accept a `std::optional<NullPlacement>` instead of NullPlacement, which did lead to some changes in downstream APIs and bindings. ~I also need some help with fixing the `c_glib` bindings.~
* GitHub Issue: #38558

Light-City committed:
Fix: Add support for null sort option per sort key.

1.Reconstruct the SortKey structure and add NullPlacement.

2.Remove NullPlacement from SortOptions

3.Fix selectk not displaying non-empty results in null AtEnd scenario.

When limit k is greater than the actual table data and the table contains Null/NaN, the data cannot be obtained and only non-empty results are available.
Therefore, we support returning non-null and supporting the order of setting Null for each SortKey.

4.Add relevant unit tests and change the interface implemented by multiple versions

Light-City committed:
fix test

Light-City committed:
fix sortkey assert

Light-City committed:
fix serde test

taepper committed:
better api

taepper committed:
Merge remote-tracking branch 'refs/remotes/fork/main' into GH-38558

# Conflicts:
#	c_glib/arrow-glib/compute.cpp
#	c_glib/arrow-glib/compute.h
#	cpp/src/arrow/compute/kernels/vector_rank.cc
#	cpp/src/arrow/compute/kernels/vector_select_k.cc
#	cpp/src/arrow/compute/kernels/vector_sort.cc
#	cpp/src/arrow/compute/kernels/vector_sort_internal.h
#	python/pyarrow/_acero.pyx
#	python/pyarrow/_compute.pyx
#	python/pyarrow/array.pxi
#	python/pyarrow/tests/test_compute.py
#	python/pyarrow/tests/test_table.py

taepper committed:
Merge branch 'refs/heads/better-api' into GH-38558

# Conflicts:
#	cpp/src/arrow/compute/api_vector.cc
#	cpp/src/arrow/compute/api_vector.h
#	cpp/src/arrow/compute/kernels/vector_rank.cc
#	cpp/src/arrow/compute/kernels/vector_select_k.cc
#	cpp/src/arrow/compute/kernels/vector_sort.cc
#	cpp/src/arrow/compute/kernels/vector_sort_internal.h
#	cpp/src/arrow/compute/kernels/vector_sort_test.cc
#	cpp/src/arrow/compute/ordering.cc
#	cpp/src/arrow/compute/ordering.h

taepper committed:
merge follow-up

taepper committed:
merge follow-up

taepper committed:
formatting

taepper committed:
formatting

taepper committed:
missing unwrap_null_placement in python

taepper committed:
do not remove demoted null_placement from python api

taepper committed:
fix member name in hash_aggregate_test

taepper committed:
update ToString method output

taepper committed:
format remove extra empty line

taepper committed:
fix python interface

taepper committed:
python formatting

taepper committed:
do not pass None to python C binding

taepper committed:
format python

taepper committed:
fix minor python api mistakes

taepper committed:
python formatting

taepper committed:
do not rename member of exposed API struct

taepper committed:
do not rename member of exposed API struct, missed test file

taepper committed:
format cc file

taepper committed:
updating api that was not using additional argument for some reason (most likely human-error while merging)

taepper committed:
make null_placement optional in the python acero api

taepper committed:
minor additional fixes

taepper committed:
amend c_glib api to use std::optional<NullPlacement> for RankOptions

taepper committed:
amend c_glib api to use std::optional<NullPlacement> for RankOptions

taepper committed:
amend c_glib api to use std::optional<NullPlacement> for RankOptions

taepper committed:
amend c_glib api to use std::optional<NullPlacement> for RankOptions

taepper committed:
fix mistake where default null placement was AtStart

taepper committed:
fix mistake where null_placement was not correctly set when sort_keys are empty

taepper committed:
fix c_glib bindings

taepper committed:
formatting

taepper committed:
formatting

Taepper commented:
Thank you for your comments!

I indeed had the same considerations. Why I opted for option 1 was indeed the clear path forward for the post-deprecation API.

Maybe option 2 is worth more consideration for the missing API breakage.

Should the `SelectKOptions` (which did not have a `NullPlacement` before) receive an additional global fallback, or should rather implement their fallback internally and non-configurable?

pitrou commented:
Well, the API breakage isn't critical either IMHO. I'd just like to have opinions from other core developers. @lidavidm Perhaps?

lidavidm commented:
As long as we're still doing major releases, then I think the slight breakage is acceptable in return for a nicer API in the end.

Taepper commented:
Alright, also note that this tries to be analogous to the deprecation of `skip_nulls` in `SetLookupOptions`:

https://github.com/apache/arrow/pull/36739/files#diff-6bc7ecec6a4f7bcefc2511cde3bd809340ad0d94bb8f7cc5f4994063c798f2faR313

pitrou commented:
> Alright, also note that this tries to be analogous to the deprecation of `skip_nulls` in `SetLookupOptions`:
> 
> https://github.com/apache/arrow/pull/36739/files#diff-6bc7ecec6a4f7bcefc2511cde3bd809340ad0d94bb8f7cc5f4994063c798f2faR313

That's a good comparison point, thank you.

zanmato1984 commented:
I prefer a conceptually simpler model and sacrifice the API compatibility:

> 1. (As done in this PR) Add a required `NullPlacement` in `SortKey`, and make the `NullPlacement` in `SortOptions` optional and deprecated

pitrou commented:
Ok, it seems everyone agrees with this approach, so let's go for it.

taepper committed:
Merge remote-tracking branch 'origin/main' into GH-38558

# Conflicts:
#	cpp/src/arrow/compute/api_vector.h

taepper committed:
Merge branch 'main' into GH-38558

taepper committed:
also update RankQuantileOptions to new NullPlacement api

taepper committed:
update library version documentation

taepper committed:
rename GARROW_OPTIONAL_NULL_PLACEMENT_UNSET to GARROW_OPTIONAL_NULL_PLACEMENT_UNSPECIFIED

taepper committed:
formatting

taepper committed:
ruby default value formatting

taepper committed:
change red-arrow gem to use different syntax for specifying null_placement

taepper committed:
update ruby test functions that use direct GObject introspection

taepper committed:
simplify helper functions

taepper committed:
move and export helper functions

taepper committed:
fixup move and export helper functions

taepper committed:
format ruby

taepper committed:
fix another ruby test

taepper committed:
fix another ruby test

taepper committed:
Improve `GArrowOptionalNullPlacement` documentation

taepper committed:
Make static_cast of GARROW_OPTIONAL_NULL_PLACEMENT safer by having GARROW_OPTIONAL_NULL_PLACEMENT_UNSPECIFIED correspond to -1. All other (possibly future) values will have a 1:1 mapping

taepper committed:
improve placement of helper garrow_optional_null_placement_* helper functions in c_glib/arrow-glib/compute.{hpp,cpp}

taepper committed:
more ruby test fixes

kou reviewed (approved):
+1

taepper committed:
improve syntax of ruby test

kou commented:
I'll merge this in a few days if nobody objects it.

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -74,6 +74,120 @@ class SelectKComparator<SortOrder::Descending> {
+    int64_t null_count = p.null_count();
+    // non-null nan null
+    if (null_placement_ == NullPlacement::AtEnd) {
+      extract_non_null_count = non_null_count <= k_ ? non_null_count : k_;
```
Can you use `std::min` for better readability?

Taepper commented on python/pyarrow/_acero.pyx:
```diff
@@ -254,18 +261,19 @@ class OrderByNodeOptions(_OrderByNodeOptions):
     """
 
-    def __init__(self, sort_keys=(), *, null_placement="at_end"):
+    def __init__(self, sort_keys=(), *, null_placement=None):
```
I am not quite sure how Deprecation in the python bindings should work, but I went with this after looking at some other files:


```
def __init__(self, sort_keys=(), *, null_placement=None):
        if null_placement is not None:
            warnings.warn(
                "Specifying null_placement in OrderByNodeOptions is deprecated "
                "as of 24.0.0. Specify null_placement per sort_key instead."
            )
        self._set_options(sort_keys, null_placement)
```

taepper committed:
wip review

taepper committed:
wip review

taepper committed:
minor renaming

taepper committed:
fixup tests

taepper committed:
deprecate one function

taepper committed:
fix select_k doc text

taepper committed:
revert old pr changes to vector_select_k.cc

taepper committed:
another select_k_test

taepper committed:
version almost finished. Some second key issues

taepper committed:
improvements

taepper committed:
fixes

taepper committed:
one more comment for later

taepper committed:
move from source_range to source_count to make helper viable for chunked arrays

taepper committed:
prepare ChunkedArray implementation

taepper committed:
simple TableSelector fix

taepper committed:
clean up TableSelector

taepper committed:
improve helper functions for clearer logic

taepper committed:
fix helper function

taepper committed:
fix array function

taepper committed:
cleaner helper function

taepper committed:
fixed ChunkedArraySelector

diff --git c/cpp/src/arrow/compute/kernels/vector_select_k.cc i/cpp/src/arrow/compute/kernels/vector_select_k.cc
index 8c14abd07e..ed1a89b027 100644
--- c/cpp/src/arrow/compute/kernels/vector_select_k.cc
+++ i/cpp/src/arrow/compute/kernels/vector_select_k.cc
@@ -16,6 +16,7 @@
 // under the License.

 #include <algorithm>
+#include <queue>
 #include <span>

 #include "arrow/compute/function.h"
@@ -82,8 +83,9 @@ struct OutputRangesByNullLikeness {
 };

 OutputRangesByNullLikeness calculateNumberNonNullAndNullLikesToTake(
-    int64_t non_null_like_count, int64_t nan_count, int64_t null_count, int64_t k,
-    NullPlacement null_placement, uint64_t* output_begin) {
+    int64_t non_null_like_count, int64_t nan_count, int64_t null_count,
+    NullPlacement null_placement, std::span<uint64_t> output_indices) {
+  int64_t k = output_indices.size();
   int64_t non_null_like_to_take = 0;
   int64_t nan_to_take = 0;
   int64_t null_to_take = 0;
@@ -91,38 +93,31 @@ OutputRangesByNullLikeness calculateNumberNonNullAndNullLikesToTake(
     non_null_like_to_take = std::min(k, non_null_like_count);
     nan_to_take = std::min(k - non_null_like_to_take, nan_count);
     null_to_take = std::min(k - non_null_like_to_take - nan_to_take, null_count);
-    // TODO.TAE make this prettier
     return OutputRangesByNullLikeness{
-        .non_null_like_output = {output_begin, output_begin + non_null_like_to_take},
-        .nan_output = {output_begin + non_null_like_to_take,
-                       output_begin + non_null_like_to_take + nan_to_take},
-        .null_output = {
-            output_begin + non_null_like_to_take + nan_to_take,
-            output_begin + non_null_like_to_take + nan_to_take + null_to_take}};
+        .non_null_like_output = output_indices.subspan(0, non_null_like_to_take),
+        .nan_output = output_indices.subspan(non_null_like_to_take, nan_to_take),
+        .null_output =
+            output_indices.subspan(non_null_like_to_take + nan_to_take, null_to_take)};
   } else {
     null_to_take = std::min(k, null_count);
     nan_to_take = std::min(k - null_to_take, nan_count);
     non_null_like_to_take = std::min(k - null_to_take - nan_to_take, non_null_like_count);
-    // TODO.TAE make this prettier
     return OutputRangesByNullLikeness{
-        .non_null_like_output = {output_begin + null_to_take + nan_to_take,
-                                 output_begin + null_to_take + nan_to_take +
-                                     non_null_like_to_take},
-        .nan_output = {output_begin + null_to_take,
-                       output_begin + null_to_take + nan_to_take},
-        .null_output = {output_begin, output_begin + null_to_take}};
+        .non_null_like_output =
+            output_indices.subspan(null_to_take + nan_to_take, non_null_like_to_take),
+        .nan_output = output_indices.subspan(null_to_take, nan_to_take),
+        .null_output = output_indices.subspan(0, null_to_take)};
   }
 }

 template <typename Comparator>
 void HeapSortNonNullsToOutput(std::span<uint64_t> non_null_input_range, Comparator cmp,
                               std::span<uint64_t> output_range) {
-  std::span<uint64_t> heap{non_null_input_range.begin(),
-                           non_null_input_range.begin() + output_range.size()};
+  std::span<uint64_t> heap = non_null_input_range.subspan(0, output_range.size());
   std::make_heap(heap.begin(), heap.end(), cmp);
-  for (auto iter = non_null_input_range.begin() + output_range.size();
-       iter != non_null_input_range.end(); ++iter) {
-    uint64_t x_index = *iter;
+
+  std::span<uint64_t> remaining_input = non_null_input_range.subspan(output_range.size());
+  for (uint64_t x_index : remaining_input) {
     if (cmp(x_index, heap.front())) {
       std::pop_heap(heap.begin(), heap.end(), cmp);
       heap.back() = x_index;
@@ -132,14 +127,12 @@ void HeapSortNonNullsToOutput(std::span<uint64_t> non_null_input_range, Comparat

   // fill output in reverse when destructing,
   // as the "worst" (next-to-would-have-been-replaced) element is at heap-top
-  // TODO.TAE remove these &*
-  uint64_t* heap_begin = &*heap.begin();
-  uint64_t* heap_end = &*heap.begin() + output_range.size();
   for (auto reverse_out_iter = output_range.rbegin();
        reverse_out_iter != output_range.rend(); reverse_out_iter++) {
-    *reverse_out_iter = *heap_begin;  // heap-top has the next element
-    std::pop_heap(heap_begin, heap_end, cmp);
-    --heap_end;
+    *reverse_out_iter = heap.front();  // heap-top has the next element
+    std::ranges::pop_heap(heap, cmp);
+    // Decrease heap-size by one
+    heap = heap.first(heap.size() - 1);
   }
 }

@@ -158,7 +151,6 @@ void HeapSortNonNullsToOutput(std::span<uint64_t> non_null_input_range,
 }

 template <typename InType>
-// TODO.TAE Could merge l and output into one span now
 void HeapSortNonNullsToOutput(std::span<uint64_t> non_null_input_range,
                               const typename TypeTraits<InType>::ArrayType& arr,
                               SortOrder order, std::span<uint64_t> output_range) {
@@ -171,6 +163,28 @@ void HeapSortNonNullsToOutput(std::span<uint64_t> non_null_input_range,
   }
 }

+struct NullNanPartitionResult {
+  std::span<uint64_t> non_null_like_range;
+  std::span<uint64_t> null_range;
+  std::span<uint64_t> nan_range;
+};
+
+template <typename ArrayType, typename Partitioner>
+NullNanPartitionResult PartitionNullsAndNans(uint64_t* indices_begin,
+                                             uint64_t* indices_end,
+                                             const ArrayType& values, int64_t offset,
+                                             NullPlacement null_placement) {
+  // Partition nulls at start (resp. end), and null-like values just before (resp. after)
+  NullPartitionResult p = PartitionNullsOnly<Partitioner>(indices_begin, indices_end,
+                                                          values, offset, null_placement);
+  NullPartitionResult q = PartitionNullLikes<ArrayType, Partitioner>(
+      p.non_nulls_begin, p.non_nulls_end, values, offset, null_placement);
+  return NullNanPartitionResult{
+      .non_null_like_range = {q.non_nulls_begin, q.non_nulls_end},
+      .null_range = {p.nulls_begin, p.nulls_end},
+      .nan_range = {q.nulls_begin, q.nulls_end}};
+}
+
 class ArraySelector : public TypeVisitor {
  public:
   ArraySelector(ExecContext* ctx, const Array& array, const SelectKOptions& options,
@@ -220,25 +234,22 @@ class ArraySelector : public TypeVisitor {
     auto* output = take_indices->template GetMutableValues<uint64_t>(1);

     // From k, calculate
-    //   l = non-null elements to take from PartitionResult
-    //   m = null-like elements to take from PartitionResult
-    // k = l + m if enough elements in input
+    //   l = non_null_like elements to take from PartitionResult
+    //   m = nan elements to take from PartitionResult
+    //   n = null elements to take from PartitionResult
+    // k = l + m + n because k was clipped to arr.length()
+    // And directly compute the ranges in {output, output+k} where we will need to place
+    // the selected elements from each group -> no longer need to track null_placement
     auto output_ranges = calculateNumberNonNullAndNullLikesToTake(
         static_cast<int64_t>(p.non_nulls_end - p.non_nulls_begin),
-        0,  // TODO.TAE it would be okay to consider these equal, but better not?
-        static_cast<int64_t>(p.nulls_end - p.nulls_begin), k_, null_placement_, output);
+        0,  // TODO.TAE it would be okay to consider null/nan equal, but better not?
+        static_cast<int64_t>(p.nulls_end - p.nulls_begin), null_placement_,
+        {output, output + k_});

-    if (null_placement_ == NullPlacement::AtEnd) {
-      HeapSortNonNullsToOutput<InType, sort_order>(
-          {p.non_nulls_begin, p.non_nulls_end}, arr, output_ranges.non_null_like_output);
-      std::copy(p.nulls_begin, p.nulls_begin + output_ranges.null_output.size(),
-                output_ranges.null_output.begin());
-    } else {
-      std::copy(p.nulls_begin, p.nulls_begin + output_ranges.null_output.size(),
-                output_ranges.null_output.begin());
-      HeapSortNonNullsToOutput<InType, sort_order>(
-          {p.non_nulls_begin, p.non_nulls_end}, arr, output_ranges.non_null_like_output);
-    }
+    HeapSortNonNullsToOutput<InType, sort_order>({p.non_nulls_begin, p.non_nulls_end},
+                                                 arr, output_ranges.non_null_like_output);
+    std::copy(p.nulls_begin, p.nulls_begin + output_ranges.null_output.size(),
+              output_ranges.null_output.begin());

     *output_ = Datum(take_indices);
     return Status::OK();
@@ -290,25 +301,24 @@ class ChunkedArraySelector : public TypeVisitor {
   VISIT_SORTABLE_PHYSICAL_TYPES(VISIT)
 #undef VISIT

-  //  template<typename InType>
-  //  int64_t ComputeNanCount(){
-  //    using GetView = GetViewType<InType>;
-  //    using ArrayType = typename TypeTraits<InType>::ArrayType;
-  //    if constexpr (has_null_like_values<typename ArrayType::TypeClass>()) {
-  //      int64_t nan_count = 0;
-  //      for (const auto& chunk : physical_chunks_) {
-  //        auto values = std::make_shared<ArrayType>(chunk->data());
-  //        int64_t length = values->length();
-  //        for(int64_t index = 0; index < length; ++index){
-  //          if(std::isnan(values->GetView(index))){
-  //            nan_count++;
-  //          }
-  //        }
-  //      }
-  //      return nan_count;
-  //    }
-  //    return 0;
-  //  }
+  template <typename InType>
+  int64_t ComputeNanCount() {
+    using ArrayType = typename TypeTraits<InType>::ArrayType;
+    if constexpr (has_null_like_values<typename ArrayType::TypeClass>()) {
+      int64_t nan_count = 0;
+      for (const auto& chunk : physical_chunks_) {
+        auto values = std::make_shared<ArrayType>(chunk->data());
+        int64_t length = values->length();
+        for (int64_t index = 0; index < length; ++index) {
+          if (std::isnan(values->GetView(index))) {
+            nan_count++;
+          }
+        }
+      }
+      return nan_count;
+    }
+    return 0;
+  }

   template <typename InType, SortOrder sort_order>
   Status SelectKthInternal() {
@@ -323,14 +333,28 @@ class ChunkedArraySelector : public TypeVisitor {
     if (k_ > chunked_array_.length()) {
       k_ = chunked_array_.length();
     }
-    //    int64_t null_count = chunked_array_.null_count();
-    //    int64_t nan_count = ComputeNanCount<InType>();
-    // TODO.TAE    int64_t non_null_like_count = chunked_array_.length() - null_count -
-    // nan_count;
+
+    ARROW_ASSIGN_OR_RAISE(auto take_indices,
+                          MakeMutableUInt64Array(k_, ctx_->memory_pool()));
+    auto* output_begin = take_indices->GetMutableValues<uint64_t>(1);
+
+    int64_t null_count = chunked_array_.null_count();
+    int64_t nan_count = ComputeNanCount<InType>();
+    int64_t non_null_like_count = chunked_array_.length() - null_count - nan_count;
+
+    auto output = calculateNumberNonNullAndNullLikesToTake(
+        non_null_like_count, nan_count, null_count, null_placement_,
+        {output_begin, output_begin + k_});
+
+    // Now we can independently fill the output with non_null, nan and null items.
+    // For non_null, we do a heap_sort, the others can just be copied until
+    // nan_taken = output.nan_range.size() and
+    // null_taken = output.null_range.size() respectively
+    size_t nan_taken = 0;
+    size_t null_taken = 0;

     std::function<bool(const HeapItem&, const HeapItem&)> cmp;
     SelectKComparator<sort_order> comparator;
-
     cmp = [&comparator](const HeapItem& left, const HeapItem& right) -> bool {
       const auto lval = GetView::LogicalValue(left.array->GetView(left.index));
       const auto rval = GetView::LogicalValue(right.array->GetView(right.index));
@@ -338,9 +362,9 @@ class ChunkedArraySelector : public TypeVisitor {
     };
     using HeapContainer =
         std::priority_queue<HeapItem, std::vector<HeapItem>, decltype(cmp)>;
-
     HeapContainer heap(cmp);
     std::vector<std::shared_ptr<ArrayType>> chunks_holder;
+
     uint64_t offset = 0;
     for (const auto& chunk : physical_chunks_) {
       if (chunk->length() == 0) continue;
@@ -352,12 +376,29 @@ class ChunkedArraySelector : public TypeVisitor {
       uint64_t* indices_end = indices_begin + indices.size();
       std::iota(indices_begin, indices_end, 0);

-      auto kth_begin = std::min(indices_begin + k_, indices_end);
-      uint64_t* iter = indices_begin;
-      for (; iter != kth_begin && heap.size() < static_cast<size_t>(k_); ++iter) {
+      const auto p = PartitionNullsAndNans<ArrayType, NonStablePartitioner>(
+          indices_begin, indices_end, arr, 0, null_placement_);
+
+      // First do nulls and nans
+      auto iter = p.null_range.begin();
+      for (; iter != p.null_range.end() && null_taken < output.null_output.size();
+           ++iter) {
+        output.null_output[null_taken] = offset + *iter;
+        null_taken++;
+      }
+      iter = p.nan_range.begin();
+      for (; iter != p.nan_range.end() && nan_taken < output.nan_output.size(); ++iter) {
+        output.nan_output[nan_taken] = offset + *iter;
+        nan_taken++;
+      }
+
+      iter = p.non_null_like_range.begin();
+      for (; iter != p.non_null_like_range.end() &&
+             heap.size() < output.non_null_like_output.size();
+           ++iter) {
         heap.push(HeapItem{*iter, offset, &arr});
       }
-      for (; iter != indices_end && !heap.empty(); ++iter) {
+      for (; iter != p.non_null_like_range.end() && !heap.empty(); ++iter) {
         uint64_t x_index = *iter;
         const auto& xval = GetView::LogicalValue(arr.GetView(x_index));
         auto top_item = heap.top();
@@ -371,16 +412,17 @@ class ChunkedArraySelector : public TypeVisitor {
       offset += chunk->length();
     }

-    auto out_size = static_cast<int64_t>(heap.size());
-    ARROW_ASSIGN_OR_RAISE(auto take_indices,
-                          MakeMutableUInt64Array(out_size, ctx_->memory_pool()));
-    auto* out_cbegin = take_indices->GetMutableValues<uint64_t>(1) + out_size - 1;
-    while (heap.size() > 0) {
-      auto top_item = heap.top();
-      *out_cbegin = top_item.index + top_item.offset;
+    // We sized output.non_null_like_output to hold exactly sufficient indices,
+    // so the heap must have been completely filled
+    assert(heap.size() == output.non_null_like_output.size());
+
+    for (auto reverse_out_iter = output.non_null_like_output.rbegin();
+         reverse_out_iter != output.non_null_like_output.rend(); reverse_out_iter++) {
+      *reverse_out_iter =
+          heap.top().index + heap.top().offset;  // heap-top has the next element
       heap.pop();
-      --out_cbegin;
     }
+
     *output_ = Datum(take_indices);
     return Status::OK();
   }
@@ -395,28 +437,6 @@ class ChunkedArraySelector : public TypeVisitor {
   Datum* output_;
 };

-struct NullNanPartitionResult {
-  std::span<uint64_t> non_null_like_range;
-  std::span<uint64_t> null_range;
-  std::span<uint64_t> nan_range;
-};
-
-template <typename ArrayType, typename Partitioner>
-NullNanPartitionResult PartitionNullsAndNans(uint64_t* indices_begin,
-                                             uint64_t* indices_end,
-                                             const ArrayType& values, int64_t offset,
-                                             NullPlacement null_placement) {
-  // Partition nulls at start (resp. end), and null-like values just before (resp. after)
-  NullPartitionResult p = PartitionNullsOnly<Partitioner>(indices_begin, indices_end,
-                                                          values, offset, null_placement);
-  NullPartitionResult q = PartitionNullLikes<ArrayType, Partitioner>(
-      p.non_nulls_begin, p.non_nulls_end, values, offset, null_placement);
-  return NullNanPartitionResult{
-      .non_null_like_range = {q.non_nulls_begin, q.non_nulls_end},
-      .null_range = {p.nulls_begin, p.nulls_end},
-      .nan_range = {q.nulls_begin, q.nulls_end}};
-}
-
 class RecordBatchSelector {
  private:
   using ResolvedSortKey = ResolvedRecordBatchSortKey;
@@ -455,13 +475,11 @@ class RecordBatchSelector {
   class SelectKForKey : public TypeVisitor {
    public:
     SelectKForKey(RecordBatchSelector* selector, size_t start_sort_key_index,
-                  std::span<uint64_t> input_indices, int64_t k_remaining,
-                  uint64_t* output_indices)
+                  std::span<uint64_t> input_indices, std::span<uint64_t> output_indices)
         : TypeVisitor(),
           selector_(selector),
           start_sort_key_index_(start_sort_key_index),
           input_indices_(input_indices),
-          k_remaining_(k_remaining),
           output_indices_(output_indices) {}

    private:
@@ -471,23 +489,24 @@ class RecordBatchSelector {
       const auto& first_remaining_sort_key = selector_->sort_keys_[start_sort_key_index_];
       const auto& arr = checked_cast<const ArrayType&>(first_remaining_sort_key.array);

-      // TODO.TAE uhh this could be prettier
-      uint64_t* input_indices_begin = &*input_indices_.begin();
-      uint64_t* input_indices_end = input_indices_begin + input_indices_.size();
+      uint64_t* input_indices_begin = input_indices_.data();
+      uint64_t* input_indices_end = input_indices_.data() + input_indices_.size();

       const auto p = PartitionNullsAndNans<ArrayType, NonStablePartitioner>(
           input_indices_begin, input_indices_end, arr, 0,
           first_remaining_sort_key.null_placement);

-      // From k, calculate
+      // From k = output_range.size(), calculate
       //   l = non_null elements to take from PartitionResult
-      //   m = null elements to take from PartitionResult
-      // k = l + m because k was clipped to num_rows()
-
+      //   m = nan elements to take from PartitionResult
+      //   n = null elements to take from PartitionResult
+      // k = l + m + n because k was clipped to num_rows()
+      // And directly compute the ranges in output_indices_ where we will need to place
+      // the selected elements from each group -> no longer need to track null_placement
       auto output_ranges = calculateNumberNonNullAndNullLikesToTake(
           static_cast<int64_t>(p.non_null_like_range.size()),
           static_cast<int64_t>(p.nan_range.size()),
-          static_cast<int64_t>(p.null_range.size()), k_remaining_,
+          static_cast<int64_t>(p.null_range.size()),
           first_remaining_sort_key.null_placement, output_indices_);

       bool last_sort_key = start_sort_key_index_ + 1 == selector_->sort_keys_.size();
@@ -496,7 +515,6 @@ class RecordBatchSelector {
         if (!output_ranges.non_null_like_output.empty()) {
           HeapSortNonNullsToOutput<InType>(p.non_null_like_range, arr,
                                            first_remaining_sort_key.order,
-                                           // TODO.TAE remove this &*
                                            output_ranges.non_null_like_output);
         }
         if (output_ranges.nan_output.size() > 0) {
@@ -521,15 +539,11 @@ class RecordBatchSelector {
         }
         if (output_ranges.nan_output.size() > 0) {
           ARROW_RETURN_NOT_OK(selector_->DoSelectKForKey(
-              start_sort_key_index_ + 1, p.nan_range, output_ranges.nan_output.size(),
-              // TODO.TAE remove this &*
-              &*output_ranges.nan_output.begin()));
+              start_sort_key_index_ + 1, p.nan_range, output_ranges.nan_output));
         }
         if (output_ranges.null_output.size() > 0) {
           ARROW_RETURN_NOT_OK(selector_->DoSelectKForKey(
-              start_sort_key_index_ + 1, p.null_range, output_ranges.null_output.size(),
-              // TODO.TAE remove this &*
-              &*output_ranges.null_output.begin()));
+              start_sort_key_index_ + 1, p.null_range, output_ranges.null_output));
         }
       }

@@ -545,14 +559,12 @@ class RecordBatchSelector {
     RecordBatchSelector* selector_;
     size_t start_sort_key_index_;
     std::span<uint64_t> input_indices_;
-    int64_t k_remaining_;
-    uint64_t* output_indices_;
+    std::span<uint64_t> output_indices_;
   };

   Status DoSelectKForKey(size_t start_sort_key_index, std::span<uint64_t> input_indices,
-                         int64_t k_remaining, uint64_t* output_indices) {
-    SelectKForKey tmp(this, start_sort_key_index, input_indices, k_remaining,
-                      output_indices);
+                         std::span<uint64_t> output_indices) {
+    SelectKForKey tmp(this, start_sort_key_index, input_indices, output_indices);
     return sort_keys_.at(start_sort_key_index).type->Accept(&tmp);
   }

@@ -571,7 +583,8 @@ class RecordBatchSelector {
     auto* output_indices = take_indices->template GetMutableValues<uint64_t>(1);

     std::span<uint64_t> input_indices_span(input_indices);
-    ARROW_RETURN_NOT_OK(DoSelectKForKey(0, input_indices_span, k_, output_indices));
+    ARROW_RETURN_NOT_OK(
+        DoSelectKForKey(0, input_indices_span, {output_indices, output_indices + k_}));
     *output_ = Datum(take_indices);
     return arrow::Status::OK();
   }

taepper committed:
fix test and add comment

taepper committed:
improve naming

taepper committed:
supress deprecation warning

taepper committed:
improve heap functions for clarity

taepper committed:
consistent style for clipping k_ to Datum length

taepper committed:
deprecate more null_placement options

taepper committed:
remove superfluous comment

Taepper commented:
@pitrou 
I reworked the implementation in `vector_select_k.cc` to be easier to understand

taepper committed:
add check against empty output span

taepper committed:
formatting

Taepper commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -74,6 +74,120 @@ class SelectKComparator<SortOrder::Descending> {
+    if (null_placement_ == NullPlacement::AtEnd) {
+      if (counter.extract_non_null_count) {
+        auto* out_cbegin = take_indices->template GetMutableValues<uint64_t>(1) +
+                           counter.extract_non_null_count - 1;
```
I improved the reverse iterator variable names in the refactored heap destruction:

```
  // fill output in reverse when destructing,
  // as the "worst" (next-to-would-have-been-replaced) element is at heap-top
  for (auto& reverse_out_iter : std::ranges::reverse_view(output_range)) {
    reverse_out_iter = heap.front();  // heap-top has the next element
    std::ranges::pop_heap(heap, cmp);
    // Decrease heap-size by one
    heap = heap.first(heap.size() - 1);
  }
```

taepper commented:
Should I remove the changes to `select_k` and only keep the interface changes in this PR?

kou commented:
@pitrou Do you want to review this?

pitrou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -102,10 +103,15 @@ class ARROW_EXPORT ArraySortOptions : public FunctionOptions {
   NullPlacement null_placement;
 };
 
+ARROW_SUPPRESS_DEPRECATION_WARNING
```
Is it necessary to wrap the entire `class` declaration, or can this macro only apply to `GetSortKeys`?

taepper replied:
I was able to move it inside but needed to manually add this block:
```
  ARROW_SUPPRESS_DEPRECATION_WARNING
  SortOptions(SortOptions&&) = default;
  SortOptions(const SortOptions&) = default;
  SortOptions& operator=(SortOptions&&) = default;
  SortOptions& operator=(const SortOptions&) = default;
  ARROW_UNSUPPRESS_DEPRECATION_WARNING
```

because the `null_placement` member is deprecated

pitrou commented on cpp/src/arrow/compute/api_vector.h:
```diff
@@ -175,43 +204,87 @@ class ARROW_EXPORT RankOptions : public FunctionOptions {
   /// Column key(s) to order by and how to order by these sort keys.
   std::vector<SortKey> sort_keys;
+
+  // DEPRECATED(set null_placement in sort_keys instead)
```
Why not use `ARROW_DEPRECATED` as in other options classes?

pitrou commented on cpp/src/arrow/compute/ordering.h:
```diff
@@ -56,12 +57,16 @@ class ARROW_EXPORT SortKey : public util::EqualityComparable<SortKey> {
+  explicit Ordering(std::vector<SortKey> sort_keys) : sort_keys_(std::move(sort_keys)) {}
+
+  // DEPRECATED(will be removed after removing null_placement from Ordering)
+  Ordering(std::vector<SortKey> sort_keys, std::optional<NullPlacement> null_placement)
```
Use `ARROW_DEPRECATED` as well?

pitrou commented on python/pyarrow/_acero.pyx:
```diff
@@ -28,6 +28,7 @@ from pyarrow.includes.libarrow_acero cimport *
 from pyarrow.lib cimport (Table, pyarrow_unwrap_table, pyarrow_wrap_table,
                           RecordBatchReader)
 from pyarrow.lib import frombytes, tobytes
+import warnings
```
Can you move this standard library import above the `pyarrow` imports?

pitrou commented on python/pyarrow/_acero.pyx:
```diff
@@ -254,18 +262,24 @@ class OrderByNodeOptions(_OrderByNodeOptions):
 
-    def __init__(self, sort_keys=(), *, null_placement="at_end"):
+    def __init__(self, sort_keys=(), *, null_placement=None):
+        if null_placement is not None:
+            warnings.warn(
+                "Specifying null_placement in OrderByNodeOptions is deprecated "
+                "as of 24.0.0. Specify null_placement per sort_key instead."
+            )
```
We should make this a `FutureWarning` (instead of the default `UserWarning`), something like:

```suggestion
            warnings.warn(
                "Specifying null_placement in OrderByNodeOptions is deprecated "
                "as of 24.0.0. Specify null_placement per sort_key instead.",
                FutureWarning
            )
```

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -2256,18 +2266,24 @@ class SortOptions(_SortOptions):
+        if null_placement is not None:
+            warnings.warn(
+                "Specifying null_placement in SortOptions is deprecated "
+                "as of 24.0.0. Specify null_placement per sort_key instead."
```
Same here: make this a `FutureWarning`

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -2499,17 +2527,26 @@ class RankOptions(_RankOptions):
 
-    def __init__(self, sort_keys="ascending", *, null_placement="at_end", tiebreaker="first"):
+    def __init__(self, sort_keys="ascending", *, null_placement=None, tiebreaker="first"):
         self._set_options(sort_keys, null_placement, tiebreaker)
```
Should a non-None null placement be deprecated here?

taepper replied:
Good question, probably yes, I added the same warning from `__init__` in `SortOptions`. Is there more to do for the deprecation?

pitrou replied:
I don't think so (except fixing the advertised deprecation version, sorry!)

pitrou commented on python/pyarrow/tests/test_table.py:
```diff
@@ -3397,17 +3397,38 @@ def test_table_sort_by(cls):
+    ], names=["a", "b", "c"])
+
+    sorted_rb = rb.sort_by([("a", "descending", "at_end"),
+                           ("b", "descending", "at_end")])
```
Can these two sort keys use different null placements, to make sure parameter unwrapping works as expected?

pitrou commented on python/pyarrow/tests/test_table.py:
```diff
@@ -3397,17 +3397,38 @@ def test_table_sort_by(cls):
+    rb = pa.RecordBatch.from_arrays([
+        pa.array([7, 35, 7, 5], type=pa.int64()),
+        pa.array([4, 1, 3, 2], type=pa.int64()),
+        pa.array(["foo", "car", "bar", "foobar"])
```
Let's make sure these columns have nulls as well.

pitrou commented on python/pyarrow/tests/test_table.py:
```diff
@@ -3397,17 +3397,38 @@ def test_table_sort_by(cls):
+    assert sorted_rb_dict["b"] == [1, 4, 3, 2]
+    assert sorted_rb_dict["c"] == ["car", "foo", "bar", "foobar"]
+
+    sorted_rb = rb.sort_by([("a", "ascending", "at_end"), ("b", "ascending", "at_end")])
```
Let's also exercise different per-column sort orders?

pitrou commented on cpp/src/arrow/compute/api_vector.cc:
```diff
@@ -138,8 +140,7 @@ static auto kArraySortOptionsType = GetFunctionOptionsType<ArraySortOptions>(
 static auto kSortOptionsType = GetFunctionOptionsType<SortOptions>(
-    DataMember("sort_keys", &SortOptions::sort_keys),
-    DataMember("null_placement", &SortOptions::null_placement));
+    CoercedDataMember("sort_keys", &SortOptions::sort_keys, &SortOptions::GetSortKeys));
```
Can you add some new examples in https://github.com/apache/arrow/blob/4f3b0f01b51c8d596334027f939fa97f45b5ac05/cpp/src/arrow/compute/function_test.cc#L45 to make sure that all SortOptions / SortKey fields are considered for comparison and serialization?

taepper replied:
Very good point. Additional tests added in `eea17e5aa3a7e42ef923dec20567d6d600096ff3` did indeed fail. They were fixed with `80bf5d29ea5b591e74fb19994e7c6dbebb10dbb2`. Is this the correct approach and is there something I might have missed?

pitrou replied:
That looks like it indeed. I don't think this is exercised much apart from unit tests TBH.

pitrou commented on cpp/src/arrow/compute/kernels/vector_sort_test.cc:
```diff
@@ -1252,9 +1248,39 @@ TEST_F(TestRecordBatchSortIndices, Null) {
+  AssertSortIndices(batch, options, "[3, 0, 5, 1, 4, 2, 6]");
+}
+
+TEST_F(TestRecordBatchSortIndices, MixedNullOrdering) {
```
Is it the exact same test as above? If it is, then perhaps remove it?

pitrou commented:
@Taepper I get this error when compiling with gcc 15.2.0:
```
In file included from /home/antoine/arrow/dev/cpp/src/arrow/compute/api.h:33,
                 from /home/antoine/arrow/dev/cpp/src/arrow/array/array_dict.cc:33:
/home/antoine/arrow/dev/cpp/src/arrow/compute/api_vector.h: In member function 'arrow::compute::Ordering arrow::compute::SortOptions::AsOrdering() &&':
/home/antoine/arrow/dev/cpp/src/arrow/compute/api_vector.h:123:58: error: converting to 'arrow::compute::Ordering' from initializer list would use explicit constructor 'arrow::compute::Ordering::Ordering(std::vector<arrow::compute::SortKey>)'
  123 |   Ordering AsOrdering() && { return {std::move(sort_keys)}; }
      |                                                          ^
In file included from /home/antoine/arrow/dev/cpp/src/arrow/compute/api_vector.h:24:
/home/antoine/arrow/dev/cpp/src/arrow/compute/ordering.h:66:12: note: 'arrow::compute::Ordering::Ordering(std::vector<arrow::compute::SortKey>)' declared here
   66 |   explicit Ordering(std::vector<SortKey> sort_keys) : sort_keys_(std::move(sort_keys)) {}
      |            ^~~~~~~~
/home/antoine/arrow/dev/cpp/src/arrow/compute/api_vector.h: In member function 'arrow::compute::Ordering arrow::compute::SortOptions::AsOrdering() const &':
/home/antoine/arrow/dev/cpp/src/arrow/compute/api_vector.h:124:51: error: converting to 'arrow::compute::Ordering' from initializer list would use explicit constructor 'arrow::compute::Ordering::Ordering(std::vector<arrow::compute::SortKey>)'
  124 |   Ordering AsOrdering() const& { return {sort_keys}; }
      |                                                   ^
/home/antoine/arrow/dev/cpp/src/arrow/compute/ordering.h:66:12: note: 'arrow::compute::Ordering::Ordering(std::vector<arrow::compute::SortKey>)' declared here
   66 |   explicit Ordering(std::vector<SortKey> sort_keys) : sort_keys_(std::move(sort_keys)) {}
      |            ^~~~~~~~
```

pitrou commented on cpp/src/arrow/compute/kernels/select_k_test.cc:
```diff
@@ -283,6 +380,62 @@ TYPED_TEST(TestSelectKWithChunkedArray, RandomValuesWithSlices) {
+  std::vector<SortKey> sort_keys{SortKey("a", SortOrder::Descending)};
+  auto options = SelectKOptions(10, sort_keys);
+  options.sort_keys[0].null_placement = NullPlacement::AtStart;
+  // These check that nulls and Nan are sorted in a stable way, but do we want that?
```
I don't think so? Same problem in `TestSelectKWithArray` by the way.

taepper replied:
I removed CheckIndices entirely

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -74,6 +77,118 @@ class SelectKComparator<SortOrder::Descending> {
+  }
+}
+
+struct PartitionResultByNullLikeness {
```
Not for this PR, but I think it would good to consolidate this into `NullPartitionResult` instead of always having to write dedicated code for null-like values.

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -102,51 +218,45 @@ class ArraySelector : public TypeVisitor {
+    //   n = null elements to take from PartitionResult
+    // k = l + m + n because k was clipped to arr.length()
+    // And directly compute the ranges in {output, output+k} where we will need to place
+    // the selected elements from each group -> no longer need to track null_placement
```
I don't think it's useful to refer to the previous implementation.

```suggestion
    // And directly compute the ranges in {output, output+k} where we will need to place
    // the selected elements from each group.
```

taepper replied:
Ah, I added this comment not as a reference to previous implementation. I wanted to signal, that all code _after_ the statement `CalculateOutputRangesByNullLikeness` is completely independent whether null_placement is `AtStart` or `AtEnd`.

So with `no longer` I wanted to refer the structuring of the code within the current implementation, not the implementation over time

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -192,6 +307,25 @@ class ChunkedArraySelector : public TypeVisitor {
+        auto values = std::make_shared<ArrayType>(chunk->data());
+        int64_t length = values->length();
+        for (int64_t index = 0; index < length; ++index) {
+          if (std::isnan(values->GetView(index))) {
```
We need to skip null values here, as a null value could have a NaN in its data slot, even though it's unlikely. You might call `IsValid`, or use the faster `VisitArraySpanInline`.

taepper replied:
Good catch. I didn't know that abstraction, thank you!

I went with this:
```
        auto visit_null = []() {};
        using T = typename GetViewType<InType>::T;
        auto visit_value = [&](T value) {
          if (std::isnan(value)) {
            nan_count++;
          }
        };
        VisitArraySpanInline<InType>(*chunk->data(), visit_value, visit_null);
```

taepper replied:
Superfluous as this function was removed when changing the calculation of `nan_count`

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -205,19 +339,38 @@ class ChunkedArraySelector : public TypeVisitor {
+    // For non_null, we do a heap_sort, the others can just be copied until
+    // nan_taken == output.nan_range.size() and
+    // null_taken == output.null_range.size() respectively
+    size_t nan_taken = 0;
+    size_t null_taken = 0;
```
Can we use `int64_t` as for other counts and lengths?

taepper replied:
I see your point for consistency, but they are only used in one single code section:

```
      for (; iter != p.null_range.end() && null_taken < output.null_range.size();
           ++iter) {
        output.null_range[null_taken] = offset + *iter;
        null_taken++;
      }
```

and this would introduce a required `static_cast`, which makes this choice a little questionable

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -252,16 +418,17 @@ class ChunkedArraySelector : public TypeVisitor {
-      *out_cbegin = top_item.index + top_item.offset;
+    // We sized output.non_null_like_range to hold exactly sufficient indices,
+    // so the heap must have been completely filled
+    assert(heap.size() == output.non_null_like_range.size());
```
Nit: use `DCHECK_EQ`

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -167,6 +278,9 @@ struct TypedHeapItem {
 
 class ChunkedArraySelector : public TypeVisitor {
+  using ResolvedSortKey = ResolvedTableSortKey;
+  using Comparator = MultipleKeyComparator<ResolvedSortKey>;
```
Are these appropriate in `ChunkedArraySelector`?

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -205,19 +339,38 @@ class ChunkedArraySelector : public TypeVisitor {
+    int64_t null_count = chunked_array_.null_count();
+    int64_t nan_count = ComputeNanCount<InType>();
+    int64_t non_null_like_count = chunked_array_.length() - null_count - nan_count;
+
```
Instead of having to separately compute the NaN count, can we call `PartitionNullsAndNans` on all chunks in advance, and then just sum the resulting NaN counts?

taepper replied:
That sounds like a better way! I will refactor the method

taepper replied:
This indeed worked out quite nicely! The temporary storage of all `indices` for the individual chunks could still be a bit better. Maybe I should introduce a struct that holds all three of `array`, `indices` and `partitions`?

And I hope naming the individual parts of a `ChunkedArray` `chunks` and not `arrays` is consistent. At least other code I checked also referred to the parts as `chunks`

pitrou replied:
> The temporary storage of all `indices` for the individual chunks could still be a bit better. Maybe I should introduce a struct that holds all three of `array`, `indices` and `partitions`?

I don't think we care in this PR. Otherwise we have `CompressedChunkLocation` that can make things more compact.

pitrou reviewed (commented):
Sorry for the delay @Taepper . This is looking very good, here are a bunch of comments.

taepper committed:
fix compilation errors and correctly place deprecations and suppresions

taepper committed:
correctly specify deprecations in python bindings

taepper committed:
remove duplicated test

taepper committed:
improve coverage in test_record_batch_sort()

taepper committed:
update select_k doc

taepper committed:
do not check for stable sort when interface is not guaranteed stable

taepper committed:
add tests for SortOption serialization and equality

taepper committed:
add all fields of SortKey to serialization

taepper committed:
refactor ChunkedArraySelector and address minor comments

taepper commented:
Thank you for the review. I was not sure regarding the introduction of `std::span` usage, which I find very succinct.

For example, also this `std::ranges::` based for-loop replacement has now been detected by my local `clangd`:
```
for (uint64_t& reverse_out_iter : std::ranges::reverse_view(output.non_null_like_range))
{...}
```

instead of

```
for (auto reverse_out_iter = output.non_null_like_range.rbegin();
     reverse_out_iter != output.non_null_like_range.rend(); reverse_out_iter++) {
```

Feel free to let me know if these code-style preferences are not wanted

taepper commented:
Of course I will rebase onto main and update the code deprecation references to 25.0.0 from 24.0.0, but did no want to introduce noise to the history for now. When you are okay with the approach, I will rebase

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -315,73 +473,124 @@ class RecordBatchSelector : public TypeVisitor {
+          input_indices_begin, input_indices_end, arr, 0,
+          first_remaining_sort_key.null_placement);
+
+      // From k = output_range.size(), calculate
```
Nit
```suggestion
      // From k = output_indices.size(), calculate
```

pitrou commented:
> Thank you for the review. I was not sure regarding the introduction of `std::span` usage, which I find very succinct.
> 
> For example, also this `std::ranges::` based for-loop replacement has now been detected by my local `clangd`:

Oh, yes, these are nice additions, thank you.

pitrou commented:
> Of course I will rebase onto main and update the code deprecation references to 25.0.0 from 24.0.0, but did no want to introduce noise to the history for now. When you are okay with the approach, I will rebase

Actually, please rebase now, I think it will make local testing easier.

taepper committed:
Merge remote-tracking branch 'origin/main' into GH-38558

taepper committed:
update variable name in comment

taepper committed:
update deprecation warnings and `since` doc-strings to arrow 25.0.0

taepper commented:
> > Of course I will rebase onto main and update the code deprecation references to 25.0.0 from 24.0.0, but did no want to introduce noise to the history for now. When you are okay with the approach, I will rebase
> 
> Actually, please rebase now, I think it will make local testing easier.

Okay, the merge actually had no conflicts, so my wariness was not necessary

taepper committed:
fix test result of added test

taepper committed:
suppress null_placement deprecation warnings in arrow-glib

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -205,40 +317,94 @@ class ChunkedArraySelector : public TypeVisitor {
+    for (const auto& chunk : physical_chunks_) {
+      if (chunk->length() == 0) continue;
+      chunks_holder.emplace_back(std::make_shared<ArrayType>(chunk->data()));
+      ArrayType& arr = *chunks_holder.back();
```
Let's please make this `const`?

pitrou commented on cpp/src/arrow/compute/kernels/vector_select_k.cc:
```diff
@@ -205,40 +317,94 @@ class ChunkedArraySelector : public TypeVisitor {
+    DCHECK_EQ(chunks_holder.size(), partitions_by_chunk.size());
+    uint64_t offset = 0;
+    for (size_t chunk_id = 0; chunk_id < chunks_holder.size(); ++chunk_id) {
+      ArrayType& arr = *chunks_holder[chunk_id];
```
Make this `const` too.

pitrou reviewed (approved):
LGTM, just two nits. Assuming CI is green, we'll be able to merge this.

pitrou committed:
Nit: make array const
